File: frontend/lib/data/models/operations.dart
class Shift {
  final String id;
  final String tenantId;
  final String name;
  final String startTime;
  final String endTime;
  final int gracePeriodMinutes;
  final DateTime createdAt;
  final DateTime updatedAt;

  Shift({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.startTime,
    required this.endTime,
    required this.gracePeriodMinutes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Shift.fromJson(Map<String, dynamic> json) {
    return Shift(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      startTime: json['start_time'] as String,
      endTime: json['end_time'] as String,
      gracePeriodMinutes: json['grace_period_minutes'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class Timesheet {
  final String id;
  final String tenantId;
  final String employeeId;
  final DateTime date;
  final String taskDescription;
  final double hoursWorked;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Join
  final String? employeeName;

  Timesheet({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.date,
    required this.taskDescription,
    required this.hoursWorked,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.employeeName,
  });

  factory Timesheet.fromJson(Map<String, dynamic> json) {
    return Timesheet(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      date: DateTime.parse(json['date'] as String),
      taskDescription: json['task_description'] as String? ?? '',
      hoursWorked: (json['hours_worked'] as num?)?.toDouble() ?? 0.0,
      status: json['status'] as String? ?? 'pending',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      employeeName: json['employee_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'tenant_id': tenantId,
        'employee_id': employeeId,
        'date': date.toIso8601String(),
        'task_description': taskDescription,
        'hours_worked': hoursWorked,
        'status': status,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt.toIso8601String(),
      };
}

class Project {
  final String id;
  final String tenantId;
  final String name;
  final String clientName;
  final String description;
  final DateTime startDate;
  final DateTime? endDate;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;

  Project({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.clientName,
    required this.description,
    required this.startDate,
    this.endDate,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      clientName: json['client_name'] as String? ?? 'Internal',
      description: json['description'] as String? ?? '',
      startDate: DateTime.parse(json['start_date'] as String),
      endDate: json['end_date'] != null ? DateTime.parse(json['end_date'] as String) : null,
      status: json['status'] as String? ?? 'active',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class OvertimeRequest {
  final String id;
  final String tenantId;
  final String employeeId;
  final DateTime date;
  final double hoursRequested;
  final String reason;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Join
  final String? employeeName;

  OvertimeRequest({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.date,
    required this.hoursRequested,
    required this.reason,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.employeeName,
  });

  factory OvertimeRequest.fromJson(Map<String, dynamic> json) {
    return OvertimeRequest(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      date: DateTime.parse(json['date'] as String),
      hoursRequested: (json['hours'] as num?)?.toDouble() ?? 0.0,
      reason: json['reason'] as String? ?? '',
      status: json['status'] as String? ?? 'pending',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      employeeName: json['employee_name'] as String?,
    );
  }
}

File: backend/internal/repository/operations.go
package repository

import (
	"context"
	"time"

	"hr-saas/internal/db"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Shift struct {
	ID                 uuid.UUID `json:"id"`
	TenantID           uuid.UUID `json:"tenant_id"`
	Name               string    `json:"name"`
	StartTime          string    `json:"start_time"` // HH:MM:SS
	EndTime            string    `json:"end_time"`   // HH:MM:SS
	GracePeriodMinutes int       `json:"grace_period_minutes"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type Timesheet struct {
	ID              uuid.UUID `json:"id"`
	TenantID        uuid.UUID `json:"tenant_id"`
	EmployeeID      uuid.UUID `json:"employee_id"`
	Date            time.Time `json:"date"`
	TaskDescription string    `json:"task_description"`
	HoursWorked     float64   `json:"hours_worked"`
	Status          string    `json:"status"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
	// Join
	EmployeeName string `json:"employee_name,omitempty"`
}

type Project struct {
	ID          uuid.UUID  `json:"id"`
	TenantID    uuid.UUID  `json:"tenant_id"`
	Name        string     `json:"name"`
	ClientName  string     `json:"client_name"`
	Description string     `json:"description"`
	StartDate   time.Time  `json:"start_date"`
	EndDate     *time.Time `json:"end_date,omitempty"`
	Status      string     `json:"status"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

type OvertimeRequest struct {
	ID           uuid.UUID  `json:"id"`
	TenantID     uuid.UUID  `json:"tenant_id"`
	EmployeeID   uuid.UUID  `json:"employee_id"`
	Date         time.Time  `json:"date"`
	Hours        float64    `json:"hours"`
	Reason       string     `json:"reason"`
	Status       string     `json:"status"`
	ApprovedByID *uuid.UUID `json:"approved_by_id,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
	// Join
	EmployeeName string `json:"employee_name,omitempty"`
}

// ---------------------------------------------------------
// SHIFTS
// ---------------------------------------------------------

func ListShifts(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Shift, error) {
	var list []Shift
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx, `SELECT id, tenant_id, name, start_time::text, end_time::text, grace_period_minutes, created_at, updated_at FROM shifts`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r Shift
			if err := rows.Scan(&r.ID, &r.TenantID, &r.Name, &r.StartTime, &r.EndTime, &r.GracePeriodMinutes, &r.CreatedAt, &r.UpdatedAt); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

// ---------------------------------------------------------
// TIMESHEETS
// ---------------------------------------------------------

func ListTimesheets(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Timesheet, error) {
	var list []Timesheet
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT t.id, t.tenant_id, t.employee_id, t.date, t.task_description, t.hours_worked, t.status, t.created_at, t.updated_at,
			        e.name as employee_name
			 FROM timesheets t
			 JOIN employees e ON t.employee_id = e.id
			 ORDER BY t.date DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r Timesheet
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.Date, &r.TaskDescription, &r.HoursWorked, &r.Status, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

// ---------------------------------------------------------
// PROJECTS
// ---------------------------------------------------------

func ListProjects(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Project, error) {
	var list []Project
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx, `SELECT id, tenant_id, name, client_name, description, start_date, end_date, status, created_at, updated_at FROM projects ORDER BY created_at DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r Project
			if err := rows.Scan(&r.ID, &r.TenantID, &r.Name, &r.ClientName, &r.Description, &r.StartDate, &r.EndDate, &r.Status, &r.CreatedAt, &r.UpdatedAt); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

func CreateProject(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, p *Project) (*Project, error) {
	var created Project
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`INSERT INTO projects (tenant_id, name, client_name, description, start_date, end_date, status)
			 VALUES ($1, $2, $3, $4, $5, $6, $7)
			 RETURNING id, tenant_id, name, client_name, description, start_date, end_date, status, created_at, updated_at`,
			tenantID, p.Name, p.ClientName, p.Description, p.StartDate, p.EndDate, p.Status,
		).Scan(&created.ID, &created.TenantID, &created.Name, &created.ClientName, &created.Description, &created.StartDate, &created.EndDate, &created.Status, &created.CreatedAt, &created.UpdatedAt)
	})
	return &created, err
}

// ---------------------------------------------------------
// OVERTIME
// ---------------------------------------------------------

func ListOvertimeRequests(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]OvertimeRequest, error) {
	var list []OvertimeRequest
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT o.id, o.tenant_id, o.employee_id, o.date, o.hours, o.reason, o.status, o.approved_by_id, o.created_at, o.updated_at,
			        e.name as employee_name
			 FROM overtime_requests o
			 JOIN employees e ON o.employee_id = e.id
			 ORDER BY o.date DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r OvertimeRequest
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.Date, &r.Hours, &r.Reason, &r.Status, &r.ApprovedByID, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

func CreateOvertimeRequest(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, o *OvertimeRequest) (*OvertimeRequest, error) {
	var created OvertimeRequest
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`INSERT INTO overtime_requests (tenant_id, employee_id, date, hours, reason, status)
			 VALUES ($1, $2, $3, $4, $5, $6)
			 RETURNING id, tenant_id, employee_id, date, hours, reason, status, approved_by_id, created_at, updated_at`,
			tenantID, o.EmployeeID, o.Date, o.Hours, o.Reason, "pending",
		).Scan(&created.ID, &created.TenantID, &created.EmployeeID, &created.Date, &created.Hours, &created.Reason, &created.Status, &created.ApprovedByID, &created.CreatedAt, &created.UpdatedAt)
	})
	return &created, err
}
